test/Constraints/fixes.swift

```diff
@@ -24,27 +24,27 @@ func f7(_ a: A, _: Int) -> Int { }
 func forgotCall() {
   // Simple cases
   var x: Int
-  x = f1 // expected-error{{function produces expected type 'Int'; did you mean to call it with '()'?}}{{9-9=()}}
+  x = f1 // expected-error{{expected an expression of type 'Int', but expression has function type returning 'Int'; did you mean to call the function with '()'?}}{{9-9=()}}
```

Comment:
Iis this correct? `expected an expression of type 'Int', but expression has function type returning 'Int'` is the same type `Int` ...

Reply:
The idea here is that we expect an `Int`, but what we got was an `()->Int`. I thought about phrasing it as `expected an expression of type 'Int', but expression has function type '()->Int'; did you mean to call the function with '()'?` instead, but this fix also applies to situations where `f1` has type `(String)->Int`, but the `String` parameter has a default value. Maybe the message should be customized in that situation though.

Reply:
Ahh I see, sorry just got confused with the message ... one suggestion maybe is to use `but expression has function type that returns the same type`. Instead of referring the type twice :)

Reply:
Maybe better if the expression is a `DeclRef` expression, tailor it to `but expression 'f1' is function that returns the same type` ... WDYT?

Reply:
I really like this idea! I made changes to customize the message if we have a specific decl we can refer to

Reply:
FWIW, `expression resolves to a {DescriptiveDeclKind}%0 {Identifier}%1 returning expected type 'Int'; did you mean to call it with '()'?` or `expression resolves to a {DescriptiveDeclKind}%0 {Identifier}%1 returning 'Int', but expected type 'Int'; did you mean to call it with '()'?`